### ISSUE TITLE
SEV replacement will not work as expected

### ISSUE THREAD

**Issue by johnwbyrd** [llvm-mos/llvm-mos#43]
To compensate for the missing SEV instruction on the 6502, a case in MOSInstrInfo.cpp has been added to try to emit a BIT instruction in place of a SEV instruction:

```
  case MOS::V:
    if (Val == 1) {
      // The V flag may be set by a BIT of a 16-bit address, which contains a
      // byte with bit 6 set.
      Builder.buildInstr(MOS::BITAbs, {Register(MOS::V)}, {Register(MOS::A)})
          .addExternalSymbol("__set_v");
      MI.eraseFromParent();
      return;
    }
```

The intention here is to use a BIT instruction to compensate.  However, there are two problems here.  First, if this code actually executes, then in debug mode an assert will occur as follows:

```
*** Bad machine code: Using an undefined physical register ***
- function:    set_entities
- basic block: %bb.2 for.body (0x29026f327b8)
- instruction: $v = BITAbs $a, &__set_v
- operand 1:   $a
```

This assert is thrown from MachineVerifier.cpp, around line 2231.

Second, this workaround does not do what it's intended to do.  The BIT instruction overwrites whatever's in the N flag with whatever's in bit 7 of the accessed memory.  Ergo, set_v trashes N unexpectedly.

Codegen does try to use V as a general purpose 1 bit flag.

Several workarounds are possible.  A software implementation of SEV is possible, but writing a version that keeps all the other flags intact might go something like:

```
php
pla
ora #$40
pha
plp
```

which seems pretty expensive just to set the V bit, and it trashes the accumulator to boot.  Better off, perhaps, not to assume the V bit is generally usable by codegen as a 1-bit register.

**Comment by johnwbyrd** [comment]
See also the 65C02's TSB, TRB, BBS, and BBR instructions.

**Comment by mysterymath** [comment]
For the use of $a, LLVM has an undef flag to mark a register as "used, but not really". I forgot to add this flag to this use of $a; I'll correct this shortly.

Regarding N and Z, I realized that so many instructions clobber them, it's not really worth modelling them explicitly except for the instructions that directly deal with them. Accordingly, we maintain as an invariant that N and Z cannot contain live values outside of certain tightly bundled patterns. Broadly, this allows us to emit pseudoinstructions without worrying about N and Z. This makes dealing with N and Z more complicated, but it's better to make something relatively rare complicated (N/Z usage) over something relatively common (basically every pseudoinstruction). I was able to take this invariant far enough that instructions no longer even bear "implicit-def $nz"; the fact that instructions clobber N and Z isn't needed by the compiler anymore, since the invariant picks up the slack.
